### PR TITLE
Create value_offset.py

### PR DESCRIPTION
**Pull request by GTaulats**
- Moves the selected strips and/or handlers as many frames or seconds as specified.
- It allows to move by seconds by taking the current framerate.
- It is flexible to the user's direction, by changing the sign of the value when going to left/right.
- It allows to enter an specific value and the strips will move to the corresponding frame.
- It could be a bit confusig to use when going left, right and changing the value manually because of the negative values, but once the user understands how it behaves, it can be really useful.

How spected to be used:

1. Press left/right shortcut.
2. Specify value type and the value itself.

To use recursively:

1. Press left/right shortcut.
2. Set a convinient fraction of seconds or a good amount of frames.
3. Use the left/right shortcut as wanted.

I have an unsolveable issue when passing seconds to frames, where some frames are not applied properly when using this operation recursively with decimal seconds not multiples of the framerate.
Ex: 30 fps and 4 additions of 0.25 s, where you end up 28 frames ahead instead of 30: 30 * 0.25=12.5~12 → 12 * 4=28 !!
It wouldn't happen for example with 60 fps: 60 * 0.25=15 → 15 * 4=60

### CODE REVIEW

**Comment by NathanLovato** [comment]
How does this enhance your editing experience over using the interactive move (G), snapping, slide, extend, and all the others?

Before adding any new tool to the add-on and to the sequencer, we have to be sure it solves an actual problem, something that no other tool solves for you. Otherwise, it tends to become another tool that nobody uses.

For example, you can move your time cursor to a point in time and snap one strip or the entire selection to it, achieving, it seems, the same effect. Maybe you won't jump exactly 0.5 seconds, but editing is all about what's happening on the image and you pretty much never move things using a time offset known in advance, but rather interactively or using the time cursor.

**Comment by NathanLovato** [comment]
Note that I can also achieve the same behavior by pressing G followed by a positive or negative number, with the added bonus that it's interactive. The only difference I see right now is offsetting in seconds over frames.

**Comment by GTaulats** [comment]
The main focus of this tool are the seconds offset, in fact I added the option with frames just this morning as it's just directly apply the number.
This is a tool I wanted the moment I saw how to advance the strips given seconds. It's time consuming for the user to calculate how many frames are x seconds depending on the framerate, and kills the flow.
This tool is not that useful for footage, but more for audio. I think to be able to control the start and end of each strip by seconds is a huge timesaver, because you can make consistent changes regardless of the framerate.
And you are right, the only difference are the seconds, as well as it's the only reason I did this tool.

**Comment by NathanLovato** [comment]
Thanks for the explanation. I'm still unsure of the tool's _necessity_ but I'll give it a try hands-on to see. I've got a lot of work so please bear with me, power sequencer's kind on the back-burner right now.

**Comment by GTaulats** [comment]
Don't worry, I understand. Check it out whenever you're finished even if it means months, as I also have other things to do😁

**Comment by NathanLovato** [comment]
I'm finally getting to this. Just tested it and it's good, I'll clean up the code and merge it. Some notes.

In the poll method, avoid checking for lists; when you do this, the engine has to populate and iterate over the list, and poll gets called often, so this can affect performance. Favor checking the active strip, even if you'd like the operator to work on the selection.

Not:

```py
    def poll(cls, context):
        return context.selected_sequences
```

Do:

```py
return context.scene.sequence_editor.active_strip
```

**Comment by NathanLovato** [comment]
When running an operator at the end of yours, you can use it in your return statement instead of `return {"FINISHED"}`

In this case:

```py
return bpy.ops.transform.seq_slide(value=(frames_offset, 0))
```

That way, if the context is wrong, the operator may cancel or what-not. This is a small detail.

**Comment by NathanLovato** [comment]
Don't overrride the object's properties when converting them. Mutating variables is a common source of bugs in software programming. Also, it can add complexity. You can simplify your code by avoiding code like this:

```py
offset_frames = round(self.value_offset)
self.value_offset = frames_offset
```

Basically, you only want to do that if you need to store the value for another run, when reusing the operator object.

**Comment by NathanLovato** [comment]
Here's the execute() function cleaned up for comparison

```py
    def execute(self, context):
        offset_frames = (
            convert_duration_to_frames(context, self.offset)
            if self.value_type == "seconds"
            else self.value
        )
        if self.direction == "left":
            offset_frames *= -1.0
        return bpy.ops.transform.seq_slide(value=(offset_frames, 0))
```

**Comment by NathanLovato** [comment]
Thanks for the contribution!

**Comment by GTaulats** [comment]
Thank you, this made my day!!

The poll() thing I actually didn't know how it worked so I just copied it from the channel_offset project, hehe.

The thing about complexity though... I did all that "mess" to be able to imput direct numbers to the value, like clicking on the seconds value, and enter positive or negative numbers, or to be able to scroll through them. That's why I used a fake direction, to set the custom value from the user.
The solution you did was indeed much easier, smarter and bug friendly, but it doesn't directly allow the seconds input (Above all negative values). Also, I purposely changed the sign of the seconds number whenever you are moving left or right. Try using the function to the left and scroll the value, you'll see what I mean. It's confusing even for me and was actually a problem I tackled when creating the tool. What I did was to allow negative values and make notice the user that the value is applied from the initial position of the strip, how like "G" works.

Summing up: The method I used allows direct interaction with the value, where you can set a negative or positive number and the strip will move relative to the initial position, in a way I think is more intuitive. Mine also allows negative values, and shows the user that the strip is moving to the left, also making the tool more intuitive. I can understand my solution may be troublesome, but meets all the features I thought convenient to have. Please consider this.

**Comment by NathanLovato** [comment]
> Mine also allows negative values, and shows the user that the strip is moving to the left, also making the tool more intuitive. I can understand my solution may be troublesome, but meets all the features I thought convenient to have. Please consider this.

Ah, sure. I'll fix that right away.

**Comment by NathanLovato** [comment]
Just so you know, there's a method `invoke()` you can use for when the *user* invokes the command initially, by pressing a keyboard shortcut, clicking a button...

```
    def invoke(self, context, event):
        return self.execute(context)
```

It's one of those things you have to know, but that would be the simplest way to affect the value based on the direction:

```py
    def invoke(self, context, event):
        self.offset = abs(self.offset)
        if self.direction == "left":
            self.offset *= -1.0
        return self.execute(context)
```

**Comment by NathanLovato** [comment]
You can see the changes in this commit: 25cc3dfbcb7013d51855790398bc6f9ce4fc85df

You can now use the slider with negative numbers again, and that shouldn't affect shortcuts or buttons.

Please let me know if the behavior is incorrect.
